src/graph/pickle.cc

```diff
@@ -14,85 +17,136 @@ namespace dgl {
 
 HeteroPickleStates HeteroPickle(HeteroGraphPtr graph) {
   HeteroPickleStates states;
-  states.metagraph = graph->meta_graph();
-  states.num_nodes_per_type = graph->NumVerticesPerType();
-  states.adjs.resize(graph->NumEdgeTypes());
+  dmlc::MemoryStringStream ofs(&states.meta);
+  dmlc::Stream *strm = &ofs;
+  strm->Write(ImmutableGraph::ToImmutable(graph->meta_graph()));
+  strm->Write(graph->NumVerticesPerType());
+  std::vector<SparseFormat> fmts;
+  std::vector<char> flags;
```

Comment:
Why does `vector<char>` needed here? You can read/write booleans separately, why you prefer first constructing a vector and then write them?